Replace node.meta source_fn with source_fn_stack

ydwu4 opened the pull request:
A resubmit of https://github.com/pytorch/pytorch/pull/108447. Copy over the descriptions:

This is a follow-up of the discussion in https://github.com/pytorch/pytorch/pull/108356, where we want to repalce source_fn with source_fn_stack

Before this PR, for the following example:
```python
backend = EagerAndRecordGraphs()

@torch.compile(backend=backend, fullgraph=True)
def cond_f(pred, pred2, x, y):
    def true_fn(pred2, x, y):
        return x + y

    def false_fn(pred2, x, y):
        def true_fn2(x, y):
            return x.sin() - y.cos()

        def false_fn2(x, y):
            return x.cos() - y.sin()

        return control_flow.cond(pred2, true_fn2, false_fn2, (x, y))

    return control_flow.cond(pred, true_fn, false_fn, (pred2, x, y))
```
The graph captured is shown below:
```python
class GraphModule(torch.nn.Module):
    def forward(self, L_pred_ : torch.Tensor, L_pred2_ : torch.Tensor, L_x_ : torch.Tensor, L_y_ : torch.Tensor):
        l_pred_ = L_pred_
        l_pred2_ = L_pred2_
        l_x_ = L_x_
        l_y_ = L_y_
        
        cond_true_1 = self.cond_true_1
        cond_false_1 = self.cond_false_1
        cond = torch.ops.higher_order.cond(l_pred_, cond_true_1, cond_false_1, [l_pred2_, l_x_, l_y_]);  l_pred_ = cond_true_1 = cond_false_1 = l_pred2_ = l_x_ = l_y_ = None
        return (cond,)
        
    class GraphModule(torch.nn.Module):
        def forward(self, l_pred2_, l_x_, l_y_):
            add = l_x_ + l_y_;  l_x_ = l_y_ = None
            return add
            
    class GraphModule(torch.nn.Module):
        def forward(self, l_pred2_, l_x_, l_y_):
            cond_true_0 = self.cond_true_0
            cond_false_0 = self.cond_false_0
            cond = torch.ops.higher_order.cond(l_pred2_, cond_true_0, cond_false_0, [l_x_, l_y_]);  l_pred2_ = cond_true_0 = cond_false_0 = l_x_ = l_y_ = None
            return cond
            
        class GraphModule(torch.nn.Module):
            def forward(self, l_x_, l_y_):
                sin = l_x_.sin();  l_x_ = None
                cos = l_y_.cos();  l_y_ = None
                sub = sin - cos;  sin = cos = None
                return sub
                
        class GraphModule(torch.nn.Module):
            def forward(self, l_x_, l_y_):
                cos = l_x_.cos();  l_x_ = None
                sin = l_y_.sin();  l_y_ = None
                sub = cos - sin;  cos = sin = None
                return sub
```
the source_fn for inner cond, sin, cos will be a (name, target) tuple:
```
('cond', <torch._ops.HigherOrderOperator object at xxx>)
('sin', 'sin')
('cos', 'cos')
('sub'. <built-in function sub>)
```

After this pr, the source_fn_stack will be a list of (name, target) tuple. The bottom of stack is the end of the list.
```
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>)], 
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>), ('sin', 'sin')], 
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cos', 'cos')]
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>), ('sub', <built-in function sub>)]
```


Test Plan:
See added tests in test_higher_order_ops.py and modify existing test.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

jerryzh168 commented:
can you also add some examples for the format for the previous "source_fn" and the current "source_fn_stack" for both (one fn stack and two fn stack)?

ydwu4 commented:
> can you also add some examples for the format for the previous "source_fn" and the current "source_fn_stack" for both (one fn stack and two fn stack) in PR summary?

Added in the PR summary

jerryzh168 commented:
> ('cond', <torch._ops.HigherOrderOperator object at xxx>)

thanks, can you add a print for the graph and subgraphs as well? and add some comments for the results?
e.g. for
`('cond', <torch._ops.HigherOrderOperator object at xxx>)`
is this (node_name, target)?

and for the source_fn_stack, why `('cond', <torch._ops.HigherOrderOperator object at xxx>)` appeared multiple times?

ydwu4 commented:
> add a print for the graph and subgraphs, and some comments for the results?
Sure! See the updated summary.

> why ('cond', <torch._ops.HigherOrderOperator object at xxx>) appeared multiple times?
I used a more complicated example where we have a nested cond and the operators will have the whole cond -> cond -> op stack as a list of (name, target) tuple.

jerryzh168 commented:
thanks! makes sense

jerryzh168 commented:
@ydwu4 for the order, I think what we have in other things is going from child to parent, e.g. (nn_module_stack), could you align with that?

ydwu4 commented:
> going from child to parent, e.g. (nn_module_stack)

No, nn_module_stack goes from parent to child see [here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/symbolic_convert.py#L1927-L1930C56) . It's an ordered_dict and the order of it is the key insertion order and the current nn_module is recorded when we calling into a child module.

jerryzh168 commented on torch/_export/serde/serialize.py:
```diff
@@ -1305,12 +1305,13 @@ def deserialize_meta_func(serialized_target: str):
-            ret["source_fn"] = (source_fn[0], op)
-
+            source_fn_st = []
+            for source_fn_str in source_fn_st_str.split(";"):
```
nit: should ";" be `st_delimiter` as well

jerryzh168 commented on torch/_export/serde/serialize.py:
```diff
@@ -459,18 +459,18 @@ def serialize_metadata(self, node: torch.fx.Node) -> Dict[str, str]:
         if stack_trace := node.meta.get("stack_trace"):
             ret["stack_trace"] = stack_trace
 
+        st_delimiter = ";"
```
looks like this is reused by deserialization, maybe it should be moved outside of the function

jerryzh168 commented on torch/_export/serde/serialize.py:
```diff
@@ -1305,12 +1305,13 @@ def deserialize_meta_func(serialized_target: str):
-
+            source_fn_st = []
+            for source_fn_str in source_fn_st_str.split(";"):
+                name, target_str = source_fn_str.split(",")
```
is serialization is using ";" here?
L469: st_delimiter.join(source_fn_list)

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -1305,12 +1305,13 @@ def deserialize_meta_func(serialized_target: str):
+            for source_fn_str in source_fn_st_str.split(";"):
+                name, target_str = source_fn_str.split(",")
+                source_fn_st.append((name, deserialize_meta_func(target_str)))
+            ret["source_fn_stack"] = source_fn_st
```
I realized, this will break bc because existing serialized models will expect a singular source fn. So can you add some casing thats like "if previously it was a single source fn do xx else do this"

ydwu4 replied:
yeah, that's a good point and many things will be broken e.g. the meta data name, etc. Maybe we should just fix these internal code and ignore the rest..  Any comments? @SherlockNoMad

jerryzh168 replied:
@angelayi could you point to the test? looks like it's not catched in OSS CI or sandcastle?

zou3519 commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -798,10 +798,11 @@ def _prepare_model_args():
                 if min_seq_nr < 0:
                     min_seq_nr = seq_nr
-                mod_name = node.meta.get("source_fn", "")
+                source_fn_stack = node.meta.get("source_fn_stack", [])
```
Some questions (maybe for @SherlockNoMad and @jerryzh168?):
- What does source_fn get used for in general?
- Why are we serializing source_fn into the exported graph? Are there use cases for this?

angelayi replied:
source_fn is used mainly for partitioning ops that are getting decomposed right now. Mainly, we'll look for all the ops that were decomposed from a torch.nn.Linear or torch.nn.functional.linear, and partition them into a subgraph.

zou3519 commented on torch/_export/serde/serialize.py:
```diff
@@ -465,12 +467,11 @@ def serialize_metadata(self, node: torch.fx.Node) -> Dict[str, str]:
                 for k, v in nn_module_stack.items()
             ]
-            ret["nn_module_stack"] = ";".join(nn_module_list)
+            ret["nn_module_stack"] = ST_DELIMITER.join(nn_module_list)
```
since we are technically changing the IR, does ExportedProgram's version number need a bump? Or is the version number bump automatic (and in-line with the PyTorch version number?)

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -150,6 +150,7 @@ def speculate_subgraph(
     graph_checkpoint,
     checkpoint,
     description,
+    source_target,
```
can we put this after the `*`? Also, what is the type of this? (HigherOrderOpVariable ?)

ydwu4 replied:
Sure! It's the .value of HigherOrderOpVariable, which is the target of the proxy that we created for the higherOrderOperator.

zou3519 replied:
Oh, so this is just a string. The name "source_target" is not that descriptive, could we add a comment with what you said?

ydwu4 replied:
Sounds good!

zou3519 reviewed (approved):
seems reasonable, some questions, otherwise minor comments

ydwu4 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1692,6 +1692,147 @@ def fn(x):
+            "add": ["wrap", "add"],
+            "linear": ["wrap", "wrap", "linear"],
+        }
+        self._check_source_fn_stack(gm, exp_stack)
```
You mean turning the list ["wrap", "cos"] into a string and expected inline the string, right? We can do that

zou3519 replied:
Yeah. you can use pprint.pprint to stringify it nicely

ydwu4 committed:
Replace node.meta source_fn with source_fn_stack (#108595)

Summary:
X-link: https://github.com/pytorch/executorch/pull/210

A resubmit of https://github.com/pytorch/pytorch/pull/108447. Copy over the descriptions:

This is a follow-up of the discussion in https://github.com/pytorch/pytorch/pull/108356, where we want to repalce source_fn with source_fn_stack

Before this PR, for the following example:
```python
backend = EagerAndRecordGraphs()

torch.compile(backend=backend, fullgraph=True)
def cond_f(pred, pred2, x, y):
    def true_fn(pred2, x, y):
        return x + y

    def false_fn(pred2, x, y):
        def true_fn2(x, y):
            return x.sin() - y.cos()

        def false_fn2(x, y):
            return x.cos() - y.sin()

        return control_flow.cond(pred2, true_fn2, false_fn2, (x, y))

    return control_flow.cond(pred, true_fn, false_fn, (pred2, x, y))
```
The graph captured is shown below:
```python
class GraphModule(torch.nn.Module):
    def forward(self, L_pred_ : torch.Tensor, L_pred2_ : torch.Tensor, L_x_ : torch.Tensor, L_y_ : torch.Tensor):
        l_pred_ = L_pred_
        l_pred2_ = L_pred2_
        l_x_ = L_x_
        l_y_ = L_y_

        cond_true_1 = self.cond_true_1
        cond_false_1 = self.cond_false_1
        cond = torch.ops.higher_order.cond(l_pred_, cond_true_1, cond_false_1, [l_pred2_, l_x_, l_y_]);  l_pred_ = cond_true_1 = cond_false_1 = l_pred2_ = l_x_ = l_y_ = None
        return (cond,)

    class GraphModule(torch.nn.Module):
        def forward(self, l_pred2_, l_x_, l_y_):
            add = l_x_ + l_y_;  l_x_ = l_y_ = None
            return add

    class GraphModule(torch.nn.Module):
        def forward(self, l_pred2_, l_x_, l_y_):
            cond_true_0 = self.cond_true_0
            cond_false_0 = self.cond_false_0
            cond = torch.ops.higher_order.cond(l_pred2_, cond_true_0, cond_false_0, [l_x_, l_y_]);  l_pred2_ = cond_true_0 = cond_false_0 = l_x_ = l_y_ = None
            return cond

        class GraphModule(torch.nn.Module):
            def forward(self, l_x_, l_y_):
                sin = l_x_.sin();  l_x_ = None
                cos = l_y_.cos();  l_y_ = None
                sub = sin - cos;  sin = cos = None
                return sub

        class GraphModule(torch.nn.Module):
            def forward(self, l_x_, l_y_):
                cos = l_x_.cos();  l_x_ = None
                sin = l_y_.sin();  l_y_ = None
                sub = cos - sin;  cos = sin = None
                return sub
```
the source_fn for inner cond, sin, cos will be a (name, target) tuple:
```
('cond', <torch._ops.HigherOrderOperator object at xxx>)
('sin', 'sin')
('cos', 'cos')
('sub'. <built-in function sub>)
```

After this pr, the source_fn_stack will be a list of (name, target) tuple. The bottom of stack is the end of the list.
```
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>)],
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>), ('sin', 'sin')],
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cos', 'cos')]
[('cond', <torch._ops.HigherOrderOperator object at xxx>), ('cond', <torch._ops.HigherOrderOperator object at xxx>), ('sub', <built-in function sub>)]
```


Test Plan:
See added tests in test_higher_order_ops.py and modify existing test.
Also updated bin by running:
"buck2 run @//mode/dev-nosan fbcode//aibench/api:gen_test_files --config client.id=nuclide"

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Reviewed By: angelayi

Differential Revision: D48984986

Pulled By: ydwu4
